File: module3/class1/example2-m3p1e2-concurrent-io-problems-solved.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <thread>
#include <mutex>
#include <shared_mutex>
#include <chrono>
#include <random>
#include <atomic>
#include <sstream>
#include <conio.h>  // For _kbhit() on Windows
#include <cstdio>   // For remove()
#include <condition_variable>

// ====================================================================
// CONFIGURATION VARIABLES - PROPER SYNCHRONIZATION ENABLED
// ====================================================================
const int NUM_THREADS = 6;                     // Number of concurrent threads
const int OPERATIONS_PER_THREAD = 20;          // Operations each thread performs
const int FILE_SIZE_KB = 5;                    // Size of each file in KB
const std::string SHARED_FILE = "shared_resource_safe.txt";
const std::string LOG_FILE = "concurrent_operations_safe.log";
const std::string BASE_FILENAME = "concurrent_file_safe_";
const int DELAY_BETWEEN_OPS_MS = 50;           // Delay between operations
// ====================================================================

class ConcurrentIOProblemsSolved {
private:
    // SOLUTION 1: Use atomic counters for thread-safe counting
    std::atomic<int> operationCounter{0};
    std::atomic<int> errorCounter{0};
    std::atomic<long long> totalBytesProcessed{0};
    
    // SOLUTION 2: Use mutexes for proper synchronization
    std::mutex logMutex;           // For thread-safe logging
    std::mutex sharedFileMutex;    // For shared file access
    std::shared_mutex fileLockMutex; // For reader-writer file access
    std::mutex fileOperationMutex; // For file creation/deletion coordination
    
    // SOLUTION 3: Use condition variables for proper coordination
    std::condition_variable fileReadyCondition;
    std::mutex fileReadyMutex;
    std::atomic<bool> fileReady{false};
    
    std::chrono::high_resolution_clock::time_point startTime;
    
    // Generate content for files
    std::string generateFileContent(int threadId, int operationId) {
        std::stringstream ss;
        ss << "=== CONCURRENT I/O OPERATION (SAFE VERSION) ===\n";
        ss << "Thread ID: " << threadId << "\n";
        ss << "Operation: " << operationId << "\n";
        ss << "Timestamp: " << std::chrono::duration_cast<std::chrono::milliseconds>(
               std::chrono::high_resolution_clock::now().time_since_epoch()).count() << "\n";
        ss << "Process ID: " << std::this_thread::get_id() << "\n";
        ss << std::string(50, '=') << "\n";
        
        // Fill to desired size
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(65, 90);
        
        int currentSize = ss.str().length();
        int targetSize = FILE_SIZE_KB * 1024;
        
        for (int i = currentSize; i < targetSize; ++i) {
            if (i % 80 == 79) {
                ss << '\n';
            } else {
                ss << static_cast<char>(dis(gen));
            }
        }
        
        return ss.str();
    }
    
    // SOLUTION FOR PROBLEM 2: Thread-safe shared file access
    void demonstrateSharedFileContentionSafe(int threadId) {
        for (int op = 0; op < OPERATIONS_PER_THREAD; ++op) {
            try {
                std::string content = "Thread " + std::to_string(threadId) + 
                                    " Operation " + std::to_string(op) + 
                                    " Time: " + std::to_string(std::chrono::duration_cast<std::chrono::milliseconds>(
                                        std::chrono::high_resolution_clock::now().time_since_epoch()).count()) + "\n";
                
                // SOLUTION: Use mutex to synchronize access to shared file
                {
                    std::lock_guard<std::mutex> lock(sharedFileMutex);
                    std::ofstream file(SHARED_FILE, std::ios::app);
                    if (file.is_open()) {
                        file << content;
                        file.flush();
                        file.close();
                        std::cout << "[THREAD " << threadId << "] SAFE WRITE to " << SHARED_FILE << " (Op " << op << ")" << std::endl;
                    }
                }
                
                totalBytesProcessed += content.length();
                
            } catch (const std::exception& e) {
                errorCounter++;
                safeLogging(threadId, "Error in shared file operation: " + std::string(e.what()));
            }
            
            std::this_thread::sleep_for(std::chrono::milliseconds(DELAY_BETWEEN_OPS_MS));
        }
    }
    
    // SOLUTION FOR PROBLEM 3: Thread-safe logging
    void safeLogging(int threadId, const std::string& message) {
        // SOLUTION: Use mutex for thread-safe logging
        std::lock_guard<std::mutex> lock(logMutex);
        std::ofstream logFile(LOG_FILE, std::ios::app);
        if (logFile.is_open()) {
            logFile << "[Thread " << threadId << "] " << message << std::endl;
            logFile.close();
        }
    }
    
    // SOLUTION FOR PROBLEM 4: Thread-safe file operations with proper coordination
    void demonstrateFileRaceConditionsSafe(int threadId) {
        for (int op = 0; op < OPERATIONS_PER_THREAD; ++op) {
            std::string filename = BASE_FILENAME + std::to_string(threadId) + "_" + std::to_string(op) + ".tmp";
            
            try {
                // SOLUTION: Use atomic operations for counters
                operationCounter++;
                
                // SOLUTION: Coordinate file operations with mutex
                std::unique_lock<std::mutex> operationLock(fileOperationMutex);
                
                // Create file
                std::string content = generateFileContent(threadId, op);
                std::ofstream file(filename, std::ios::binary);
                if (file.is_open()) {
                    file.write(content.c_str(), content.length());
                    file.flush(); // Ensure data is written
                    file.close();
                    
                    std::cout << "[THREAD " << threadId << "] CREATED FILE: " << filename << " (" << content.length() << " bytes)" << std::endl;
                    safeLogging(threadId, "Created file: " + filename);
                    
                    // SOLUTION FOR PROBLEM 5: Proper synchronization for read-after-write
                    // Ensure file is completely written before reading
                    std::this_thread::sleep_for(std::chrono::milliseconds(5)); // Small delay to ensure file system consistency
                    
                    // Try to read the file we just created
                    std::ifstream readFile(filename, std::ios::binary);
                    if (readFile.is_open()) {
                        readFile.seekg(0, std::ios::end);
                        size_t fileSize = readFile.tellg();
                        readFile.close();
                        
                        totalBytesProcessed += fileSize;
                        std::cout << "[THREAD " << threadId << "] READ FILE: " << filename << " (" << fileSize << " bytes)" << std::endl;
                        safeLogging(threadId, "Read file: " + filename + " (" + std::to_string(fileSize) + " bytes)");
                    } else {
                        errorCounter++;
                        std::cout << "[THREAD " << threadId << "] ERROR: Could not read file: " << filename << std::endl;
                        safeLogging(threadId, "ERROR: Could not read file: " + filename);
                    }
                    
                    // SOLUTION FOR PROBLEM 6: Coordinated file deletion
                    // Ensure no other threads are accessing the file before deletion
                    std::ifstream checkFile(filename);
                    if (checkFile.good()) {
                        checkFile.close();
                        
                        // Small delay to ensure file handles are released
                        std::this_thread::sleep_for(std::chrono::milliseconds(2));
                        
                        if (std::remove(filename.c_str()) == 0) {
                            std::cout << "[THREAD " << threadId << "] DELETED FILE: " << filename << std::endl;
                            safeLogging(threadId, "Deleted file: " + filename);
                        } else {
                            std::cout << "[THREAD " << threadId << "] ERROR: Could not delete file: " << filename << std::endl;
                            safeLogging(threadId, "ERROR: Could not delete file: " + filename);
                        }
                    }
                    
                } else {
                    errorCounter++;
                    safeLogging(threadId, "ERROR: Could not create file: " + filename);
                }
                
                // Release the operation lock
                operationLock.unlock();
                
            } catch (const std::exception& e) {
                errorCounter++;
                safeLogging(threadId, "ERROR in file operations: " + std::string(e.what()));
            }
            
            std::this_thread::sleep_for(std::chrono::milliseconds(DELAY_BETWEEN_OPS_MS));
        }
    }
    
    // SOLUTION FOR PROBLEM 7: Proper reader-writer synchronization
    void demonstrateFileLockingProblemsSafe(int threadId) {
        std::string sharedDataFile = "shared_data_safe_" + std::to_string(threadId % 3) + ".dat";
        
        for (int op = 0; op < OPERATIONS_PER_THREAD / 2; ++op) {
            try {
                if (op % 2 == 0) {
                    // Writer thread - use exclusive lock
                    std::string data = "Data from thread " + std::to_string(threadId) + 
                                     " operation " + std::to_string(op) + "\n";
                    
                    // SOLUTION: Use exclusive lock for writers
                    std::unique_lock<std::shared_mutex> writeLock(fileLockMutex);
                    std::ofstream file(sharedDataFile, std::ios::app);
                    if (file.is_open()) {
                        file << data;
                        file.flush();
                        file.close();
                        std::cout << "[THREAD " << threadId << "] SAFE WRITE to " << sharedDataFile << " (Op " << op << ")" << std::endl;
                    }
                    writeLock.unlock();
                    
                    safeLogging(threadId, "Wrote to shared file: " + sharedDataFile);
                    
                } else {
                    // Reader thread - use shared lock
                    // SOLUTION: Use shared lock for readers (multiple readers can access simultaneously)
                    std::shared_lock<std::shared_mutex> readLock(fileLockMutex);
                    std::ifstream file(sharedDataFile);
                    if (file.is_open()) {
                        std::string line;
                        int lineCount = 0;
                        while (std::getline(file, line)) {
                            lineCount++;
                        }
                        file.close();
                        
                        std::cout << "[THREAD " << threadId << "] SAFE READ from " << sharedDataFile << " (" << lineCount << " lines)" << std::endl;
                        safeLogging(threadId, "Read shared file: " + sharedDataFile + 
                                    " (" + std::to_string(lineCount) + " lines)");
                    }
                    readLock.unlock();
                }
                
            } catch (const std::exception& e) {
                errorCounter++;
                safeLogging(threadId, "ERROR in file locking demo: " + std::string(e.what()));
            }
            
            std::this_thread::sleep_for(std::chrono::milliseconds(DELAY_BETWEEN_OPS_MS));
        }
    }
    
public:
    ConcurrentIOProblemsSolved() {
        startTime = std::chrono::high_resolution_clock::now();
        
        // Clean up any existing files
        std::remove(SHARED_FILE.c_str());
        std::remove(LOG_FILE.c_str());
        
        // Clean up any existing shared data files
        for (int i = 0; i < 3; ++i) {
            std::string sharedDataFile = "shared_data_safe_" + std::to_string(i) + ".dat";
            std::remove(sharedDataFile.c_str());
        }
    }
    
    void runConcurrentOperations() {
        std::cout << "=== CONCURRENT I/O PROBLEMS - SOLVED VERSION ===" << std::endl;
        std::cout << "This program demonstrates PROPER solutions to I/O concurrency issues:" << std::endl;
        std::cout << "1. Thread-safe shared file access using mutexes" << std::endl;
        std::cout << "2. Safe logging operations with synchronization" << std::endl;
        std::cout << "3. Coordinated file creation/deletion" << std::endl;
        std::cout << "4. Reader-writer locks for file access" << std::endl;
        std::cout << "5. Atomic operations for counters" << std::endl;
        std::cout << std::string(60, '=') << std::endl;
        std::cout << "Configuration:" << std::endl;
        std::cout << "- Threads: " << NUM_THREADS << std::endl;
        std::cout << "- Operations per thread: " << OPERATIONS_PER_THREAD << std::endl;
        std::cout << "- All synchronization mechanisms: ENABLED" << std::endl;
        std::cout << std::string(60, '=') << std::endl;
        std::cout << "Press any key to stop the demonstration..." << std::endl;
        std::cout << "The program will run in continuous cycles until you press a key." << std::endl;
        std::cout << std::string(60, '-') << std::endl;
        
        int cycleCount = 0;
        bool userStopped = false;
        
        // Start monitoring for user input in a separate thread
        std::thread monitorThread([&userStopped]() {
            _getch();  // Wait for any key press
            userStopped = true;
            std::cout << "\n>>> User requested stop. Finishing current cycle..." << std::endl;
        });
        
        // Main demonstration loop
        while (!userStopped) {
            cycleCount++;
            std::cout << "\n" << std::string(60, '=') << std::endl;
            std::cout << ">>> STARTING SAFE CYCLE #" << cycleCount << " <<<" << std::endl;
            std::cout << std::string(60, '=') << std::endl;
            
            // Reset counters for this cycle
            operationCounter = 0;
            errorCounter = 0;
            totalBytesProcessed = 0;
            startTime = std::chrono::high_resolution_clock::now();
            
            std::vector<std::thread> threads;
            
            // Launch threads with proper synchronization
            std::cout << "Launching " << NUM_THREADS << " properly synchronized threads..." << std::endl;
            for (int i = 0; i < NUM_THREADS; ++i) {
                threads.emplace_back([this, i, cycleCount]() {
                    std::cout << "  Thread " << i << " started safely (Cycle " << cycleCount << ")" << std::endl;
                    
                    // Each thread performs properly synchronized I/O operations
                    demonstrateSharedFileContentionSafe(i);
                    demonstrateFileRaceConditionsSafe(i);
                    demonstrateFileLockingProblemsSafe(i);
                    
                    std::cout << "  Thread " << i << " completed safely (Cycle " << cycleCount << ")" << std::endl;
                });
            }
            
            // Wait for all threads to complete
            for (auto& thread : threads) {
                thread.join();
            }
            
            // Display results for this cycle
            std::cout << "\n" << std::string(40, '-') << std::endl;
            std::cout << "SAFE CYCLE #" << cycleCount << " RESULTS:" << std::endl;
            displayResults();
            
            if (!userStopped) {
                std::cout << "\nWaiting 3 seconds before next cycle..." << std::endl;
                std::cout << "(Press any key to stop)" << std::endl;
                
                // Wait 3 seconds or until user presses key
                for (int i = 0; i < 30 && !userStopped; ++i) {
                    std::this_thread::sleep_for(std::chrono::milliseconds(100));
                }
            }
        }
        
        // Clean up monitor thread
        if (monitorThread.joinable()) {
            monitorThread.join();
        }
        
        std::cout << "\n" << std::string(60, '=') << std::endl;
        std::cout << "SAFE DEMONSTRATION COMPLETED AFTER " << cycleCount << " CYCLES" << std::endl;
        std::cout << std::string(60, '=') << std::endl;
        std::cout << "Check the following files - they should be properly formatted:" << std::endl;
        std::cout << "- " << SHARED_FILE << " (no corruption expected)" << std::endl;
        std::cout << "- " << LOG_FILE << " (clean log entries expected)" << std::endl;
        std::cout << "- shared_data_safe_*.dat files (consistent data expected)" << std::endl;
        std::cout << "\nPress any key to exit..." << std::endl;
        _getch();
    }
    
private:
    void displayResults() {
        auto endTime = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
        
        std::cout << "\n" << std::string(60, '=') << std::endl;
        std::cout << "SAFE CONCURRENT I/O RESULTS" << std::endl;
        std::cout << std::string(60, '=') << std::endl;
        std::cout << "Execution time: " << duration.count() << " ms" << std::endl;
        std::cout << "Total threads: " << NUM_THREADS << std::endl;
        std::cout << "Expected operations: " << (NUM_THREADS * OPERATIONS_PER_THREAD) << std::endl;
        std::cout << "Actual operations: " << operationCounter.load() << std::endl;
        std::cout << "Errors encountered: " << errorCounter.load() << " (should be 0 or very low)" << std::endl;
        std::cout << "Total bytes processed: " << (totalBytesProcessed.load() / 1024.0) << " KB" << std::endl;
        
        // Analyze the shared file for corruption
        std::ifstream file(SHARED_FILE);
        if (file.is_open()) {
            std::string line;
            int lineCount = 0;
            while (std::getline(file, line)) {
                lineCount++;
            }
            file.close();
            std::cout << "Shared file lines: " << lineCount << std::endl;
        }
        
        // Analyze the log file for corruption
        std::ifstream logFile(LOG_FILE);
        if (logFile.is_open()) {
            std::string line;
            int logLines = 0;
            int wellFormedLines = 0;
            while (std::getline(logFile, line)) {
                logLines++;
                // Check for properly formatted lines
                if (line.find("[Thread") != std::string::npos) {
                    wellFormedLines++;
                }
            }
            logFile.close();
            std::cout << "Log file lines: " << logLines << std::endl;
            std::cout << "Well-formed log lines: " << wellFormedLines << " (should equal total lines)" << std::endl;
        }
        
        std::cout << std::string(60, '=') << std::endl;
        std::cout << "SAFETY ANALYSIS:" << std::endl;
        std::cout << "✓ Operations count should match expected: " << 
                    (operationCounter.load() == (NUM_THREADS * OPERATIONS_PER_THREAD) ? "PASS" : "FAIL") << std::endl;
        std::cout << "✓ Errors should be minimal: " << 
                    (errorCounter.load() <= 2 ? "PASS" : "FAIL") << std::endl;
        std::cout << "✓ All synchronization mechanisms working properly" << std::endl;
        std::cout << "✓ No race conditions expected in this version" << std::endl;
        std::cout << std::string(60, '=') << std::endl;
    }
};

int main() {
    try {
        ConcurrentIOProblemsSolved demo;
        demo.runConcurrentOperations();
    } catch (const std::exception& e) {
        std::cerr << "Fatal error: " << e.what() << std::endl;
        std::cout << "Press any key to exit..." << std::endl;
        _getch();
        return 1;
    }
    
    return 0;
}

File: module3/class1/example2-m3p1e2-concurrent-io-problems-solved.cs
using System;
using System.IO;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Diagnostics;

namespace ConcurrentIOProblemsSolved
{
    // ====================================================================
    // CONFIGURATION VARIABLES - PROPER SYNCHRONIZATION ENABLED
    // ====================================================================
    public static class IOConfig
    {
        public const int NUM_THREADS = 6;                     // Number of concurrent threads
        public const int OPERATIONS_PER_THREAD = 20;          // Operations each thread performs
        public const int FILE_SIZE_KB = 5;                    // Size of each file in KB
        public const string SHARED_FILE = "shared_resource_safe.txt";
        public const string LOG_FILE = "concurrent_operations_safe.log";
        public const string BASE_FILENAME = "concurrent_file_safe_";
        public const int DELAY_BETWEEN_OPS_MS = 50;           // Delay between operations
    }
    // ====================================================================

    class ConcurrentIOProblemsSolved
    {
        // SOLUTION 1: Use thread-safe collections and atomic operations
        private long operationCounter = 0;
        private long errorCounter = 0;
        private long totalBytesProcessed = 0;
        
        // SOLUTION 2: Use proper synchronization primitives
        private readonly object logLock = new object();           // For thread-safe logging
        private readonly object sharedFileLock = new object();    // For shared file access
        private readonly ReaderWriterLockSlim fileLockSlim = new ReaderWriterLockSlim(); // For reader-writer file access
        private readonly SemaphoreSlim fileOperationSemaphore = new SemaphoreSlim(1, 1); // For file operation coordination
        
        // SOLUTION 3: Use ManualResetEvent for coordination
        private readonly ManualResetEventSlim fileReadyEvent = new ManualResetEventSlim(false);
        
        private readonly Stopwatch stopwatch;
        private readonly Random random;
        private readonly CancellationTokenSource cancellationTokenSource;

        public ConcurrentIOProblemsSolved()
        {
            stopwatch = Stopwatch.StartNew();
            random = new Random();
            cancellationTokenSource = new CancellationTokenSource();
            
            // Clean up any existing files
            CleanupFiles();
        }

        private void CleanupFiles()
        {
            try
            {
                if (File.Exists(IOConfig.SHARED_FILE))
                    File.Delete(IOConfig.SHARED_FILE);
                if (File.Exists(IOConfig.LOG_FILE))
                    File.Delete(IOConfig.LOG_FILE);
                
                // Clean up shared data files
                for (int i = 0; i < 3; i++)
                {
                    string sharedDataFile = $"shared_data_safe_{i}.dat";
                    if (File.Exists(sharedDataFile))
                        File.Delete(sharedDataFile);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Cleanup error: {ex.Message}");
            }
        }

        private string GenerateFileContent(int threadId, int operationId)
        {
            var sb = new StringBuilder();
            sb.AppendLine("=== CONCURRENT I/O OPERATION (SAFE C# VERSION) ===");
            sb.AppendLine($"Thread ID: {threadId}");
            sb.AppendLine($"Operation: {operationId}");
            sb.AppendLine($"Timestamp: {DateTimeOffset.Now.ToUnixTimeMilliseconds()}");
            sb.AppendLine($"Managed Thread ID: {Thread.CurrentThread.ManagedThreadId}");
            sb.AppendLine(new string('=', 50));

            // Fill to desired size
            int currentSize = Encoding.UTF8.GetByteCount(sb.ToString());
            int targetSize = IOConfig.FILE_SIZE_KB * 1024;
            int remainingSize = Math.Max(0, targetSize - currentSize);

            for (int i = 0; i < remainingSize; i++)
            {
                if (i % 80 == 79)
                {
                    sb.Append('\n');
                }
                else
                {
                    sb.Append((char)random.Next(65, 91)); // A-Z characters
                }
            }

            return sb.ToString();
        }

        // SOLUTION FOR PROBLEM 2: Thread-safe shared file access
        private async Task DemonstrateSharedFileContentionSafeAsync(int threadId)
        {
            for (int op = 0; op < IOConfig.OPERATIONS_PER_THREAD; op++)
            {
                try
                {
                    string content = $"Thread {threadId} Operation {op} Time: {DateTimeOffset.Now.ToUnixTimeMilliseconds()}\n";
                    
                    // SOLUTION: Use lock to synchronize access to shared file
                    lock (sharedFileLock)
                    {
                        using (var writer = new StreamWriter(IOConfig.SHARED_FILE, append: true, Encoding.UTF8))
                        {
                            writer.Write(content);
                            writer.Flush();
                        }
                        Console.WriteLine($"[THREAD {threadId}] SAFE WRITE to {IOConfig.SHARED_FILE} (Op {op})");
                    }
                    
                    Interlocked.Add(ref totalBytesProcessed, Encoding.UTF8.GetByteCount(content));
                    
                }
                catch (Exception ex)
                {
                    Interlocked.Increment(ref errorCounter);
                    await SafeLoggingAsync(threadId, $"Error in shared file operation: {ex.Message}");
                }
                
                await Task.Delay(IOConfig.DELAY_BETWEEN_OPS_MS);
            }
        }

        // SOLUTION FOR PROBLEM 3: Thread-safe logging
        private async Task SafeLoggingAsync(int threadId, string message)
        {
            // SOLUTION: Use lock for thread-safe logging
            lock (logLock)
            {
                try
                {
                    using (var writer = new StreamWriter(IOConfig.LOG_FILE, append: true, Encoding.UTF8))
                    {
                        writer.WriteLine($"[Thread {threadId}] {message}");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Logging error: {ex.Message}");
                }
            }
        }

        // SOLUTION FOR PROBLEM 4: Thread-safe file operations with proper coordination
        private async Task DemonstrateFileRaceConditionsSafeAsync(int threadId)
        {
            for (int op = 0; op < IOConfig.OPERATIONS_PER_THREAD; op++)
            {
                string filename = $"{IOConfig.BASE_FILENAME}{threadId}_{op}.tmp";
                
                try
                {
                    // SOLUTION: Use atomic operations for counters
                    Interlocked.Increment(ref operationCounter);
                    
                    // SOLUTION: Coordinate file operations with semaphore
                    await fileOperationSemaphore.WaitAsync();
                    
                    try
                    {
                        // Create file
                        string content = GenerateFileContent(threadId, op);
                        
                        // SOLUTION FOR PROBLEM 5: Proper async file operations
                        using (var fileStream = new FileStream(filename, FileMode.Create, FileAccess.Write, FileShare.None, 4096, useAsync: true))
                        using (var writer = new StreamWriter(fileStream, Encoding.UTF8))
                        {
                            await writer.WriteAsync(content);
                            await writer.FlushAsync();
                        }
                        
                        Console.WriteLine($"[THREAD {threadId}] CREATED FILE: {filename} ({Encoding.UTF8.GetByteCount(content)} bytes)");
                        await SafeLoggingAsync(threadId, $"Created file: {filename}");
                        
                        // SOLUTION FOR PROBLEM 5: Proper synchronization for read-after-write
                        // Ensure file is completely written before reading
                        await Task.Delay(5); // Small delay to ensure file system consistency
                        
                        // Try to read the file we just created
                        if (File.Exists(filename))
                        {
                            using (var fileStream = new FileStream(filename, FileMode.Open, FileAccess.Read, FileShare.Read, 4096, useAsync: true))
                            {
                                long fileSize = fileStream.Length;
                                Interlocked.Add(ref totalBytesProcessed, fileSize);
                                
                                Console.WriteLine($"[THREAD {threadId}] READ FILE: {filename} ({fileSize} bytes)");
                                await SafeLoggingAsync(threadId, $"Read file: {filename} ({fileSize} bytes)");
                            }
                        }
                        else
                        {
                            Interlocked.Increment(ref errorCounter);
                            Console.WriteLine($"[THREAD {threadId}] ERROR: Could not read file: {filename}");
                            await SafeLoggingAsync(threadId, $"ERROR: Could not read file: {filename}");
                        }
                        
                        // SOLUTION FOR PROBLEM 6: Coordinated file deletion
                        // Ensure no other threads are accessing the file before deletion
                        await Task.Delay(2); // Small delay to ensure file handles are released
                        
                        if (File.Exists(filename))
                        {
                            try
                            {
                                File.Delete(filename);
                                Console.WriteLine($"[THREAD {threadId}] DELETED FILE: {filename}");
                                await SafeLoggingAsync(threadId, $"Deleted file: {filename}");
                            }
                            catch (Exception deleteEx)
                            {
                                Console.WriteLine($"[THREAD {threadId}] ERROR: Could not delete file: {filename} - {deleteEx.Message}");
                                await SafeLoggingAsync(threadId, $"ERROR: Could not delete file: {filename} - {deleteEx.Message}");
                            }
                        }
                    }
                    finally
                    {
                        fileOperationSemaphore.Release();
                    }
                }
                catch (Exception ex)
                {
                    Interlocked.Increment(ref errorCounter);
                    await SafeLoggingAsync(threadId, $"ERROR in file operations: {ex.Message}");
                    fileOperationSemaphore.Release();
                }
                
                await Task.Delay(IOConfig.DELAY_BETWEEN_OPS_MS);
            }
        }

        // SOLUTION FOR PROBLEM 7: Proper reader-writer synchronization
        private async Task DemonstrateFileLockingProblemsSafeAsync(int threadId)
        {
            string sharedDataFile = $"shared_data_safe_{threadId % 3}.dat";
            
            for (int op = 0; op < IOConfig.OPERATIONS_PER_THREAD / 2; op++)
            {
                try
                {
                    if (op % 2 == 0)
                    {
                        // Writer thread - use write lock
                        string data = $"Data from thread {threadId} operation {op}\n";
                        
                        // SOLUTION: Use write lock for exclusive access
                        fileLockSlim.EnterWriteLock();
                        try
                        {
                            using (var writer = new StreamWriter(sharedDataFile, append: true, Encoding.UTF8))
                            {
                                await writer.WriteAsync(data);
                                await writer.FlushAsync();
                            }
                            Console.WriteLine($"[THREAD {threadId}] SAFE WRITE to {sharedDataFile} (Op {op})");
                        }
                        finally
                        {
                            fileLockSlim.ExitWriteLock();
                        }
                        
                        await SafeLoggingAsync(threadId, $"Wrote to shared file: {sharedDataFile}");
                    }
                    else
                    {
                        // Reader thread - use read lock
                        // SOLUTION: Use read lock for shared access (multiple readers allowed)
                        fileLockSlim.EnterReadLock();
                        try
                        {
                            if (File.Exists(sharedDataFile))
                            {
                                string[] lines = await File.ReadAllLinesAsync(sharedDataFile);
                                int lineCount = lines.Length;
                                
                                Console.WriteLine($"[THREAD {threadId}] SAFE READ from {sharedDataFile} ({lineCount} lines)");
                                await SafeLoggingAsync(threadId, $"Read shared file: {sharedDataFile} ({lineCount} lines)");
                            }
                        }
                        finally
                        {
                            fileLockSlim.ExitReadLock();
                        }
                    }
                }
                catch (Exception ex)
                {
                    Interlocked.Increment(ref errorCounter);
                    await SafeLoggingAsync(threadId, $"ERROR in file locking demo: {ex.Message}");
                }
                
                await Task.Delay(IOConfig.DELAY_BETWEEN_OPS_MS);
            }
        }

        public async Task RunConcurrentOperationsAsync()
        {
            Console.WriteLine("=== CONCURRENT I/O PROBLEMS - SOLVED C# VERSION ===");
            Console.WriteLine("This program demonstrates PROPER solutions to I/O concurrency issues:");
            Console.WriteLine("1. Thread-safe shared file access using locks");
            Console.WriteLine("2. Safe logging operations with synchronization");
            Console.WriteLine("3. Coordinated file creation/deletion with semaphores");
            Console.WriteLine("4. Reader-writer locks for file access");
            Console.WriteLine("5. Atomic operations for counters");
            Console.WriteLine(new string('=', 60));
            Console.WriteLine("Configuration:");
            Console.WriteLine($"- Threads: {IOConfig.NUM_THREADS}");
            Console.WriteLine($"- Operations per thread: {IOConfig.OPERATIONS_PER_THREAD}");
            Console.WriteLine("- All synchronization mechanisms: ENABLED");
            Console.WriteLine(new string('=', 60));
            Console.WriteLine("Press any key to stop the demonstration...");
            Console.WriteLine("The program will run in continuous cycles until you press a key.");
            Console.WriteLine(new string('-', 60));

            int cycleCount = 0;
            bool userStopped = false;

            // Start monitoring for user input
            var keyTask = Task.Run(() =>
            {
                Console.ReadKey(true);
                userStopped = true;
                Console.WriteLine("\n>>> User requested stop. Finishing current cycle...");
            });

            // Main demonstration loop
            while (!userStopped)
            {
                cycleCount++;
                Console.WriteLine($"\n{new string('=', 60)}");
                Console.WriteLine($">>> STARTING SAFE CYCLE #{cycleCount} <<<");
                Console.WriteLine(new string('=', 60));

                // Reset counters for this cycle
                Interlocked.Exchange(ref operationCounter, 0);
                Interlocked.Exchange(ref errorCounter, 0);
                Interlocked.Exchange(ref totalBytesProcessed, 0);
                stopwatch.Restart();

                // Launch tasks with proper synchronization
                Console.WriteLine($"Launching {IOConfig.NUM_THREADS} properly synchronized tasks...");
                
                var tasks = new List<Task>();
                for (int i = 0; i < IOConfig.NUM_THREADS; i++)
                {
                    int threadId = i; // Capture for closure
                    tasks.Add(Task.Run(async () =>
                    {
                        Console.WriteLine($"  Task {threadId} started safely (Cycle {cycleCount})");
                        
                        // Each task performs properly synchronized I/O operations
                        await DemonstrateSharedFileContentionSafeAsync(threadId);
                        await DemonstrateFileRaceConditionsSafeAsync(threadId);
                        await DemonstrateFileLockingProblemsSafeAsync(threadId);
                        
                        Console.WriteLine($"  Task {threadId} completed safely (Cycle {cycleCount})");
                    }));
                }

                // Wait for all tasks to complete
                await Task.WhenAll(tasks);

                // Display results for this cycle
                Console.WriteLine($"\n{new string('-', 40)}");
                Console.WriteLine($"SAFE CYCLE #{cycleCount} RESULTS:");
                DisplayResults();

                if (!userStopped)
                {
                    Console.WriteLine("\nWaiting 3 seconds before next cycle...");
                    Console.WriteLine("(Press any key to stop)");
                    
                    // Wait 3 seconds or until user presses key
                    for (int i = 0; i < 30 && !userStopped; i++)
                    {
                        await Task.Delay(100);
                    }
                }
            }

            Console.WriteLine($"\n{new string('=', 60)}");
            Console.WriteLine($"SAFE DEMONSTRATION COMPLETED AFTER {cycleCount} CYCLES");
            Console.WriteLine(new string('=', 60));
            Console.WriteLine("Check the following files - they should be properly formatted:");
            Console.WriteLine($"- {IOConfig.SHARED_FILE} (no corruption expected)");
            Console.WriteLine($"- {IOConfig.LOG_FILE} (clean log entries expected)");
            Console.WriteLine("- shared_data_safe_*.dat files (consistent data expected)");
            Console.WriteLine("\nPress any key to exit...");
            Console.ReadKey(true);
        }

        private void DisplayResults()
        {
            stopwatch.Stop();
            
            Console.WriteLine($"\n{new string('=', 60)}");
            Console.WriteLine("SAFE CONCURRENT I/O RESULTS");
            Console.WriteLine(new string('=', 60));
            Console.WriteLine($"Execution time: {stopwatch.ElapsedMilliseconds} ms");
            Console.WriteLine($"Total threads: {IOConfig.NUM_THREADS}");
            Console.WriteLine($"Expected operations: {IOConfig.NUM_THREADS * IOConfig.OPERATIONS_PER_THREAD}");
            Console.WriteLine($"Actual operations: {Interlocked.Read(ref operationCounter)}");
            Console.WriteLine($"Errors encountered: {Interlocked.Read(ref errorCounter)} (should be 0 or very low)");
            Console.WriteLine($"Total bytes processed: {Interlocked.Read(ref totalBytesProcessed) / 1024.0:F2} KB");

            // Analyze the shared file for corruption
            try
            {
                if (File.Exists(IOConfig.SHARED_FILE))
                {
                    string[] lines = File.ReadAllLines(IOConfig.SHARED_FILE);
                    Console.WriteLine($"Shared file lines: {lines.Length}");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error reading shared file: {ex.Message}");
            }

            // Analyze the log file for corruption
            try
            {
                if (File.Exists(IOConfig.LOG_FILE))
                {
                    string[] lines = File.ReadAllLines(IOConfig.LOG_FILE);
                    int wellFormedLines = 0;
                    
                    foreach (string line in lines)
                    {
                        if (line.Contains("[Thread") && !string.IsNullOrWhiteSpace(line))
                        {
                            wellFormedLines++;
                        }
                    }
                    
                    Console.WriteLine($"Log file lines: {lines.Length}");
                    Console.WriteLine($"Well-formed log lines: {wellFormedLines} (should equal total lines)");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error reading log file: {ex.Message}");
            }

            Console.WriteLine(new string('=', 60));
            Console.WriteLine("SAFETY ANALYSIS:");
            
            long actualOps = Interlocked.Read(ref operationCounter);
            long expectedOps = IOConfig.NUM_THREADS * IOConfig.OPERATIONS_PER_THREAD;
            Console.WriteLine($"✓ Operations count should match expected: {(actualOps == expectedOps ? "PASS" : "FAIL")}");
            
            long errors = Interlocked.Read(ref errorCounter);
            Console.WriteLine($"✓ Errors should be minimal: {(errors <= 2 ? "PASS" : "FAIL")}");
            
            Console.WriteLine("✓ All synchronization mechanisms working properly");
            Console.WriteLine("✓ No race conditions expected in this version");
            Console.WriteLine(new string('=', 60));
        }

        public void Dispose()
        {
            fileLockSlim?.Dispose();
            fileOperationSemaphore?.Dispose();
            fileReadyEvent?.Dispose();
            cancellationTokenSource?.Dispose();
        }
    }

    class Program
    {
        static async Task Main(string[] args)
        {
            try
            {
                var demo = new ConcurrentIOProblemsSolved();
                await demo.RunConcurrentOperationsAsync();
                demo.Dispose();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Fatal error: {ex.Message}");
                Console.WriteLine("Press any key to exit...");
                Console.ReadKey(true);
            }
        }
    }
}
